README.md
# cdrcR

`cdrcR` is an R wrapper to access the CDRC APIs endpoints and retrieve CDRC data programmatically.
The package is designed to have one main function -- `getCDRC` -- which allows users to get data from all CDRC APIs endpoints.
A list of the datasets that can be accessed through these enpoints and their metadata is obtained by running `listCDRC()`. Such list will also provide users with a dataset identifier --the **dataCode**-- which will address the data request to the correct endpoint. Run `?getCDRC()` to access the function full documentation.

## Installation

You can install `cdrcR` from CRAN running:

```
install.packages("cdrcR")
```

The CRAN version currently does not retrieve OA boundaries. To get those please install the development version. Data affected are MODUM and OA classification.
You can install the development version of `cdrcR` from Github using devtools running:

```
# install.packages("devtools")
devtools::install_github("aelissa/cdrcR")

```

## Register to the CDRC

To use the CDRC APIs you need to register to the CDRC [HERE](https://apps.cdrc.ac.uk/datasetportal/Identity/Account/Register)

## Usage

Load the library.

```
library(cdrcR)
```

Log-in with the username and password that you used when registered to the CDRC.

```
loginCDRC(username="your-username",password="your-password")
```

Then you can list the datasets available and the relative dataCode which identifies the API endpoint.

```
listCDRC()

```

which will result in a data frame like the following extract:

| Title                                          | DataCode                           | dataSetURL                                                          | GeographicalCoverage | GeographyLevel |
| ---------------------------------------------- | ---------------------------------- | ------------------------------------------------------------------- | -------------------- | -------------- |
| Access to Healthy Assets & Hazards (AHAH) 2019 | AHAHInputs, AHAHOverallIndexDomain | https://data.cdrc.ac.uk/dataset/access-healthy-assets-hazards-ahah  | GreatBritain         | LSOA           |
| Classification of Workplace Zones (COWZ) 2011  | COWZUK2011                         | https://data.cdrc.ac.uk/dataset/classification-workplace-zones-cowz | UnitedKingdom        | WZ             |
| Index of Multiple Deprivation (IMD) 2019       | IMD2019                            | https://data.cdrc.ac.uk/dataset/index-multiple-deprivation-imd      | UnitedKingdom        | LSOA           |
| Internet User Classification (IUC) 2018        | IUC2018                            | https://data.cdrc.ac.uk/dataset/internet-user-classification        | GreatBritain         | LSOA           |

Pick the `DataCode` relative to the dataset you want to retrieve data about and use it as input for the dataCode parameter in `getCDRC` (run `?getCDRC()` to see the detailed documentation).

Be aware that the API endpoints enable query for the following geographies: postcodes, LSOAs, MSOAs, LAD codes and LAD names. Not all data are originally developed at these geography levels (you can find the original geography level for each dataset) with `listCDRC()`), therefore the areas that better overlap your required geography will be returned.

## Examples

1. Get the overall Access to Healthy Assets & Hazards (AHAH) index for the following postcodes: L13AY,L82TJ,L83UL
   For the Access to Healthy Assets & Hazards (AHAH) index you can chose to get either the individual inputs (via AHAHInputs) or the overall domain index (via AHAHOverallIndexDomain). In this example we are interested in the overall index to rank the postcodes above by their level of access to healthy assets and hazards. Please note that while the request is for postcodes the data is at LSOA level, therefore the LSOAs that better overlap the requested postcodes will be returned.

```
###login

loginCDRC(username="your-username",password="your-password")

###check dataCode

listCDRC()

###get the AHAH index for  postcodes L13AY,L82TJ,L83UL

ahah<-getCDRC("AHAHOverallIndexDomain",geography = "postcode", geographyCode = c("L13AY","L82TJ","L83UL"))

###rank postcodes by AHAH index from best to worst performing

ahah[order(ahah$ahah),c("postCode","ahah")]

#  postCode     ahah
#3   L8 3UL 20.01734
#2   L8 2TJ 23.04482
#1   L1 3AY 45.91745

```

2. Get the Internet User Classification across Liverpool having the LSOA codes and map it.

```
###login

loginCDRC(username="your-username",password="your-password")

###check dataCode

listCDRC()

###get Liverpool LSOAs

liverpool<-sf::st_as_sf(liverpool)

###get IUC data with geographical boundaries

iuc<-getCDRC("IUC2018",geography = "LSOA", geographyCode = liverpool$LSOA11CD, boundaries = TRUE)

###map iuc

library(ggplot2)
ggplot()+
geom_sf(aes(fill=grpLabel, group=grpLabel),iuc,show.legend = T,color="NA")+
theme_void()+
labs(fill="IUC groups")

```

2. Get the Index of Multiple Deprivation of Sheffield and Leeds Local Authority Districts.

```
###login

loginCDRC(username="your-username",password="your-password")

###check dataCode

listCDRC()

###get the IMD data for Sheffield

imd<-getCDRC("IMD2019",geography = "LADname", geographyCode = c("Sheffield","Leeds"))

```


NEWS.md
# cdrcR v0.1.0

* Updated the ONS Open Geography Portal API

# cdrcR v0.1.1

* Updated the ONS Open Geography Portal API to get Output Areas boundaries


tests/testthat.R
library(testthat)
library(CDRC)

test_check("CDRC")


R/listCDRC.R
#' This function returns a list of all CDRC datasets that can be retrieved with this package. This function takes no arguments.
#'
#' @return A dataframe of the datasets available to request with the API.
#' @examples \dontrun{
#' dataset_list <- listCDRC()
#' View(dataset_list)
#' }
#' @export
#'

listCDRC<-function(){
  token<-paste0("Bearer ",Sys.getenv("CDRC_API_KEY"))
  records <- httr::GET("https://api.cdrc.ac.uk/v1/DataSetMetaData",httr::add_headers(Authorization = token))
  if(httr::http_error(records)){
    if (records$status_code == 500){
      message("The token is expired. Login again with CDRC_login() to generate a new one.")
    }else{
      print(httr::message_for_status(records,"get the datasets list"))
      }
  }
  else{
    records<-httr::content(records)
  }

  datacode<-unnest_datacode(records)
  records<-as.data.frame(do.call(rbind,records))
  geography_level<-purrr::map(records[["granularities"]],unlist)
  geography_level<-purrr::map(geography_level,1)

  records$DataCode<-datacode
  records$GeographyLevel<-do.call(rbind,geography_level)
  records$Title<-do.call(rbind,records$title)
  records$dataSetURL<-do.call(rbind,records$dataSetURL)
  records$GeographicalCoverage<-do.call(rbind,records$spatialGeographicalCoverageLocation)
  records<-records[c("Title","DataCode","dataSetURL","GeographicalCoverage","GeographyLevel")]

  return(records)
}

unnest_datacode<-function(records){
  datacode<-list()
    for (i in 1:length(records)) {
      tmp<-purrr::map(records[[i]][["apiEndpoints"]],unlist)
      tmp<-purrr::map(tmp,"url")
      tmp<-do.call(rbind,tmp)
      tmp<-sub(".*https://api.cdrc.ac.uk/v1/", "", tmp)
      tmp<-sub("/.*", "", tmp)
      tmp<-as.data.frame(tmp)
      tmp<-paste(tmp[!(duplicated(tmp$V1) | tmp$V1=="" | tmp$V1=="https:"),],sep = "",collapse = ", ")
      datacode[[i]]<-tmp
    }
  datacode<-do.call(rbind,datacode)
  return(datacode)
}



R/loginCDRC.R
#' Login to the CDRC
#'
#' This function will log in a CDRC user based on the username and password. If you do not
#' have a CDRC user yet, please register on https://apps.cdrc.ac.uk/datasetportal/Identity/Account/Register
#' When you log in an api token is automatically generated, saved in your R environment and loaded, no further action is required from you to access the API.
#'
#' For safety reasons the API token will expire in 24h. This means that after 24h you need to log in again to generate a new API token which will be self-updated and load.
#'
#'
#' @param username A character-string with your CDRC username.
#' @param password A character-string with your CDRC password.
#' @return A response message which confirms the login was successful and how to see the API key.
#' @examples \dontrun{
#' loginCDRC(name='your-username',password='your-password')
#' }
#' @export

loginCDRC<-function(username,password){

  if(!is.character(username))stop("Username must be a string")

  request_body <- data.frame(
    username = username,
    password = password
  )

  request_body_json <- rjson::toJSON(request_body)
  result <- httr::POST("https://api.cdrc.ac.uk/Login",
                       body = request_body_json,
                       httr::accept("*/*"),
                       httr::add_headers(.headers = c("Content-Type"="application/json")))
  Output <- httr::content(result)
  print(Output[["message"]])

  save_api_key(Output[["token"]],Output[["message"]])

}


save_api_key<-function(key, message){
  home <- Sys.getenv("HOME")
  renv <- file.path(home, ".Renviron")
  if(file.exists(renv)){
    # Backup original .Renviron before doing anything else here.
    file.copy(renv, file.path(home, ".Renviron_backup"))
  }
  if(!file.exists(renv)){
    file.create(renv)
  }
  tv <- readLines(renv)
  if(!any(grepl("CDRC_API_KEY",tv))){
    keyconcat <- paste0("CDRC_API_KEY='", key, "'")
    # Append API key to .Renviron file
    write(keyconcat, renv, sep = "\n", append = TRUE)
    readRenviron("~/.Renviron")
    if(message=="Login successful"){
    message('Your API token has been stored in your .Renviron and loaded. You can now access the CDRC API! To see your token run Sys.getenv("CDRC_API_KEY").') #\nTo use now, restart R or run `readRenviron("~/.Renviron")`
    }else{message('Login must be successful to load the API token.')}
  } else {
  oldenv<-utils::read.table(renv, stringsAsFactors = FALSE)
  newenv <- oldenv[- grepl("CDRC_API_KEY", oldenv),]
  keyconcat <- paste0("CDRC_API_KEY='", key, "'")
  utils::write.table(newenv, renv, quote = FALSE, sep = "\n",
              col.names = FALSE, row.names = FALSE)
  # Append updated API key to .Renviron file
  write(keyconcat, renv, sep = "\n", append = TRUE)
  readRenviron("~/.Renviron")
  if(message=="Login successful"){
    message('Your API token has been stored in your .Renviron and loaded. You can now access the CDRC API! To see your token run Sys.getenv("CDRC_API_KEY").') #\nTo use now, restart R or run `readRenviron("~/.Renviron")`
  }else{message('Login must be successful to load the API token.')}
  }
}




R/cdrc_login.R
#' Login to the CDRC
#'
#' This function will log in a CDRC user based on the username and password. If you do not
#' have a CDRC user yet, please register on https://data.cdrc.ac.uk/user/register
#' When you log in an api token is automatically generated, saved in your R environment and loaded, no further action is required from you to access the API.
#'
#' For safety reasons the API token will expire in 24h. This means that after 24h you need to log in again to generate a new API token which will be self-updated and load.
#'
#'
#' @param username A character-string with your CDRC username.
#' @param password A character-string with your CDRC password.
#' @return A response message which confirms the login was successful and how to see the API key.
#' @examples \dontrun{
#' cdrc_login(name='your-username',password='your-password')
#' }
#' @export

cdrc_login<-function(username,password){

  if(!is.character(username))stop("Username must be a string")

  request_body <- data.frame(
    username = username,
    password = password
  )

  request_body_json <- rjson::toJSON(request_body)
  result <- httr::POST("https://api.cdrc.ac.uk/Login",
                       body = request_body_json,
                       httr::accept("*/*"),
                       httr::add_headers(.headers = c("Content-Type"="application/json")))
  Output <- httr::content(result)
  print(Output[["message"]])

  save_api_key(Output[["token"]],Output[["message"]])

}


save_api_key<-function(key, message){
  home <- Sys.getenv("HOME")
  renv <- file.path(home, ".Renviron")
  if(file.exists(renv)){
    # Backup original .Renviron before doing anything else here.
    file.copy(renv, file.path(home, ".Renviron_backup"))
  }
  if(!file.exists(renv)){
    file.create(renv)
  }
  tv <- readLines(renv)
  if(!any(grepl("CDRC_API_KEY",tv))){
    keyconcat <- paste0("CDRC_API_KEY='", key, "'")
    # Append API key to .Renviron file
    write(keyconcat, renv, sep = "\n", append = TRUE)
    readRenviron("~/.Renviron")
    if(message=="Login successful"){
    message('Your API token has been stored in your .Renviron and loaded. You can now access the CDRC API! To see your token run Sys.getenv("CDRC_API_KEY").') #\nTo use now, restart R or run `readRenviron("~/.Renviron")`
    }else{message('Login must be successful to load the API token.')}
  } else {
  oldenv<-read.table(renv, stringsAsFactors = FALSE)
  newenv <- oldenv[- grepl("CDRC_API_KEY", oldenv),]
  keyconcat <- paste0("CDRC_API_KEY='", key, "'")
  write.table(newenv, renv, quote = FALSE, sep = "\n",
              col.names = FALSE, row.names = FALSE)
  # Append updated API key to .Renviron file
  write(keyconcat, renv, sep = "\n", append = TRUE)
  readRenviron("~/.Renviron")
  if(message=="Login successful"){
    message('Your API token has been stored in your .Renviron and loaded. You can now access the CDRC API! To see your token run Sys.getenv("CDRC_API_KEY").') #\nTo use now, restart R or run `readRenviron("~/.Renviron")`
  }else{message('Login must be successful to load the API token.')}
  }
}




R/get_dataset.R
#' Get data from the CDRC
#'
#' Obtain data from the CDRC datasets. To find out what datasets are available and their respective dataCode run `list_datasets()`.
#'
#'
#' @param dataCode A character-string API identifier associated which each dataset in the CDRC. To find out the dataCode of your desired dataset run `list_datasets()`.
#' @param geography The geographical levels in which the data can be retrieved. It can be postcode, MSOA or LSOA.
#' Note that the geography in which the data are retrieved does not necessarily correspond with the geography of the data. For example, it is possible to query data of the AHAH index by postcodes although the index is originally at LSOA level. Therefore you will see returned the LSOAs that better match the required postcodes.
#' @param geographyCode A character-vector of one or more postcodes, LSOA codes or MSOA codes.
#' @param boundaries if FALSE (the default), returns a data frame of the desired data. if TRUE, uses the Open Geography Portal API to return an sf with the 'geometry' column.
#' @return A dataframe or sf depending whether boundaries are set to FALSE or TRUE respectively.
#' @examples \dontrun{
#' ahah_data <- get_dataset("AHAHOverallIndexDomain",geography = "postcode", geographyCode = c("CH430UQ","LS61EF","L83UL"), boundaries = TRUE)
#' View(ahah_data)
#' plot(ahah_data$geometry)
#' }
#' @export

get_dataset<-function(dataCode,geography=c("postcode","MSOA","LSOA"),geographyCode,boundaries=FALSE){

  geography <- match.arg(geography)
  if(any(!is.character(geographyCode)))stop("geographyCode needs to be a character vector.")
  if(geography=="MSOA"){geography="msoaCode"}
  if(geography=="LSOA"){geography="lsoaCode"}
  if(geography=="postcode"){geography="postCode"}

  data_list<-list_datasets()
  if(!any(grepl(dataCode,data_list$DataCode)))stop("The dataCode is not in the list of available datasets. Check the list with `list_datasets()`.")

  ####check geographyCode length

  if(length(geographyCode)==1){
    url <- paste0("https://api.cdrc.ac.uk/v1/",
                  dataCode,
                  "/",
                  geography,
                  "=",
                  geographyCode)
    single_code<-TRUE
  }
  else{
    single_code<-FALSE
    if(geography=="msoaCode"){
      geography<-"msoaCodes"
    }
    if(geography=="lsoaCode"){
      geography<-"lsoaCodes"
    }
    if(geography=="postCode"){
      geography<-"postCodes"
    }

    url<-paste0("https://api.cdrc.ac.uk/v1/",
                dataCode,
                "/",
                geography)
  }

  token<-paste0("Bearer ",Sys.getenv("CDRC_API_KEY"))

  if (single_code){
    call<-httr::GET(url,httr::add_headers(Authorization = token))
  } else {
    request_body_json <- rjson::toJSON(geographyCode)
    call <- httr::POST(url,
                       body = request_body_json,
                       httr::accept("*/*"),
                       httr::add_headers(Authorization = token,.headers = c("Content-Type"="application/json")))
  }
  if(httr::http_error(call)){
    if (call$status_code == 500){
      message("The token is expired. Login again with CDRC_login() to generate a new one.")
    }else{
      message(httr::message_for_status(call,"get the data"))
    }
    stop()
  }
  else{
    data<-httr::content(call,as = "text")
    data<-jsonlite::fromJSON(data)
    if(single_code==TRUE){
      data<-as.data.frame(data[[2]])
    }
    else{
      data<-as.data.frame(rlist::list.rbind(data[[2]]))
      if(any(duplicated(data[,1]))){
        data<-unique(data)
      }
    }
  }

  if(boundaries){
    data_list<-dplyr::filter(data_list, grepl(dataCode, DataCode))
    data<-get_boundaries(data,data_list$GeographyLevel,single_code)
  }

  return(data)
}

get_boundaries<-function(data,geo,single_code){
  #codeList<-paste(data[[c(1)]],sep = "",collapse = "', '")
  if(geo =="LSOA"){
    ogpURL<-"https://services1.arcgis.com/ESMARspQHYMw9BZ9/arcgis/rest/services/Lower_Layer_Super_Output_Areas_December_2011_Boundaries_EW_BFC_V2/FeatureServer/0/query"
    geocode<-"LSOA11CD"
    geocode_<-"LSOA11CD"
    geocode__<-"LSOA11CD"
  }
  if(geo =="OA"){
    ogpURL<-"https://services1.arcgis.com/ESMARspQHYMw9BZ9/arcgis/rest/services/Output_Areas_December_2011_Boundaries_EW_BGC/FeatureServer/0/query"
    geocode<-"OA11CD"
    geocode_<-"outputArea"
    geocode__<-"outputArea"
  }
  if(geo =="WZ"){
    ogpURL<-"https://ons-inspire.esriuk.com/arcgis/rest/services/Census_Boundaries/Workplace_Zone_December_2011_Boundaries/MapServer/0/query"
    geocode<-"wz11cd"
    geocode_<-"wzCode"
    geocode__<-"workPlaceZone"
  }
  cd<-data %>% dplyr::select(contains(geocode)|contains(geocode_)|contains(geocode__))

  if(nrow(cd)==1){
    whereClause<-urltools::url_encode(paste0("?where=",cd[,1]))
    ogpURL<-paste0(ogpURL,whereClause,"&outFields=*&outSR=4326&f=json")
    sf <- sf::st_read(ogpURL)
  }else{
    seq<-splitAt(cd[,1],50)
    clause<-seq %>% purrr::map(~paste(.x,sep = "",collapse = "', '"))
    clause<-clause %>% purrr::map(~paste0(geocode," IN ('",noquote(.x),"')"))
    req <- clause %>% purrr::map(~ httr::POST(url = ogpURL,
                                              body = list(where= .x,
                                              outfields="*",
                                              outSR = '4326',
                                              f='json'),
                                .headers = c("application/x-www-form-urlencoded"))
                                )
    sf <- req %>% purrr::map_df(~sf::read_sf(httr::content(.x,type='text',encoding='UTF-8')))
  }

  sf <- sf[c("geometry")]
  data<-sf::st_as_sf(cbind(data,sf))
  return(data)
}

splitAt <- function(x, pos) {
  unname(split(x,rep(c(0:(round(length(x)/pos))),each=pos)[1:length(x)]))
}











R/list_datasets.R
#' This function returns a list of all CDRC datasets that can be retrieved with this package. This function takes no arguments.
#'
#' @return A dataframe of the datasets available to request with the API.
#' @examples \dontrun{
#' dataset_list <- list_data()
#' View(dataset_list)
#' }
#' @export
#'

list_datasets<-function(){
  token<-paste0("Bearer ",Sys.getenv("CDRC_API_KEY"))
  records <- httr::GET("https://api.cdrc.ac.uk/v1/DataSetMetaData",httr::add_headers(Authorization = token))
  if(httr::http_error(records)){
    if (records$status_code == 500){
      message("The token is expired. Login again with CDRC_login() to generate a new one.")
    }else{
      print(httr::message_for_status(records,"get the datasets list"))
      }
  }
  else{
    records<-httr::content(records)
  }

  datacode<-unnest_datacode(records)
  records<-as.data.frame(do.call(rbind,records))
  geography_level<-purrr:::map(records[["granularities"]],unlist)
  geography_level<-purrr:::map(geography_level,1)

  records$DataCode<-datacode
  records$GeographyLevel<-do.call(rbind,geography_level)
  records$Title<-do.call(rbind,records$title)
  records$dataSetURL<-do.call(rbind,records$dataSetURL)
  records$GeographicalCoverage<-do.call(rbind,records$spatialGeographicalCoverageLocation)
  records<-records[c("Title","DataCode","dataSetURL","GeographicalCoverage","GeographyLevel")]

  return(records)
}

unnest_datacode<-function(records){
  datacode<-list()
    for (i in 1:length(records)) {
      tmp<-purrr:::map(records[[i]][["apiEndpoints"]],unlist)
      tmp<-purrr:::map(tmp,"url")
      tmp<-do.call(rbind,tmp)
      tmp<-sub(".*https://api.cdrc.ac.uk/v1/", "", tmp)
      tmp<-sub("/.*", "", tmp)
      tmp<-as.data.frame(tmp)
      tmp<-paste(tmp[!(duplicated(tmp$V1) | tmp$V1=="" | tmp$V1=="https:"),],sep = "",collapse = ", ")
      datacode[[i]]<-tmp
    }
  datacode<-do.call(rbind,datacode)
  return(datacode)
}



R/getCDRC.R
#' Get data from the CDRC
#'
#' Obtain data from the CDRC datasets. To find out what datasets are available and their respective dataCode run `listCDRC()`.
#'
#'
#' @param dataCode A character-string API identifier associated which each dataset in the CDRC. To find out the dataCode of your desired dataset run `listCDRC()`.
#' @param geography The geographical levels in which the data can be retrieved. It can be postcode, MSOA, LSOA, LAD or LADname.
#' Note that the geography in which the data are retrieved does not necessarily correspond with the geography of the data. For example, it is possible to query data of the AHAH index by postcodes although the index is originally at LSOA level. Therefore you will see returned the LSOAs that better match the required postcodes.
#' @param geographyCode A character-vector of one or more postcodes, LSOA codes, MSOA codes, LAD codes or LAD names.
#' @param boundaries if FALSE (the default), returns a data frame of the desired data. if TRUE, uses the Open Geography Portal API to return an sf with the 'geometry' column.
#' @return A dataframe or sf depending whether boundaries are set to FALSE or TRUE respectively.
#' @examples \dontrun{
#' ahah_data <- getCDRC("AHAHOverallIndexDomain",geography = "postcode",
#' geographyCode = c("CH430UQ","LS61EF","L83UL"), boundaries = TRUE)
#' View(ahah_data)
#' plot(ahah_data$geometry)
#' }
#' @importFrom magrittr %>%
#' @importFrom rlang :=
#' @export

getCDRC<-function(dataCode,geography=c("postcode","MSOA","LSOA","LAD","LADname"),geographyCode,boundaries=FALSE){

  geography <- match.arg(geography)
  if(any(!is.character(geographyCode)))stop("geographyCode needs to be a character vector.")


  data_list<-listCDRC()
  if(!any(grepl(dataCode,data_list$DataCode)))stop("The dataCode is not in the list of available datasets. Check the list with `listCDRC()`.")
  data_list<-dplyr::filter(data_list, grepl(dataCode, DataCode))
  if(geography==data_list$GeographyLevel){same=TRUE}else{same=FALSE}

  if(geography=="MSOA"){geography="msoaCode"}
  if(geography=="LSOA"){geography="lsoaCode"}
  if(geography=="postcode"){geography="postCode"}
  if(geography=="LAD"){geography="ladCode"}
  if(geography=="LADname"){
    geographyCode=search_ladCode(geographyCode)
    geography="ladCode"
  }

  ####check geographyCode length

  if(length(geographyCode)==1){
    url <- paste0("https://api.cdrc.ac.uk/v1/",
                  dataCode,
                  "/",
                  geography,
                  "=",
                  geographyCode)
    single_code<-TRUE
  }else{
    single_code<-FALSE
    if(geography=="msoaCode"){
      geography<-"msoaCodes"
    }
    if(geography=="lsoaCode"){
      geography<-"lsoaCodes"
    }
    if(geography=="postCode"){
      geography<-"postCodes"
    }
    if(geography=="ladCode"){
      geography<-"ladCodes"
    }
    url<-paste0("https://api.cdrc.ac.uk/v1/",
                dataCode,
                "/",
                geography)
  }

  token<-paste0("Bearer ",Sys.getenv("CDRC_API_KEY"))

  if (single_code){
    call<-httr::GET(url,httr::add_headers(Authorization = token))
  } else {
    request_body_json <- rjson::toJSON(geographyCode)
    call <- httr::POST(url,
                       body = request_body_json,
                       httr::accept("*/*"),
                       httr::add_headers(Authorization = token,.headers = c("Content-Type"="application/json")))
  }

  if(httr::http_error(call)){
    if (call$status_code == 500){
      message("The token is expired. Login again with CDRC_login() to generate a new one.")
    }else{
      message(httr::message_for_status(call,"get the data"))
    }
    stop()
  }else{
    data<-httr::content(call,as = "text")
    data<-jsonlite::fromJSON(data)
    if(single_code==TRUE){
      data<-as.data.frame(data[[2]])
    }else{
      if(same){
        data<-as.data.frame(rlist::list.rbind(data[[2]]))
      }else{
        colnames(data)[2]<-"dataCol"
        data<-tidyr::unnest(data,dataCol)
      }
    }

    if(nrow(data)==0)stop("No data found. Check the data coverage with `listCDRC()`")

    if(data_list$GeographyLevel =="LSOA"){
      geocode<-"LSOA11CD"
      geocode_<-"lsoa11"
      geocode__<-"LSOA11CD"
      init<-"OA"
    }
    if(data_list$GeographyLevel =="OA"){
      geocode<-"OA11CD"
      geocode_<-"outputArea"
      geocode__<-"outputArea"
      init<-"LS"
    }
    if(data_list$GeographyLevel =="WZ"){
      geocode<-"wz11cd"
      geocode_<-"wzCode"
      geocode__<-"workPlaceZone"
      init<-"OA"
    }

    colName<-data %>%
      dplyr::select(!tidyselect::ends_with("NM")) %>%
      dplyr::select(!tidyselect::starts_with(init))%>%
      dplyr::select(tidyselect::contains(geocode)|tidyselect::contains(geocode_)|tidyselect::contains(geocode__)) %>% colnames()
    data<-data %>%
      dplyr::rename(!!geocode := tidyselect::all_of(colName)) %>%
      as.data.frame()

    if(any(duplicated(dplyr::select(data,!!geocode)))){
      data<-data %>% dplyr::distinct(dplyr::across(tidyselect::all_of(geocode)),.keep_all = T) %>%
        as.data.frame()
    }
  }

  if(boundaries){
    data<-get_boundaries(data,data_list$GeographyLevel,single_code)
  }

  return(data)
}

get_boundaries<-function(data,geo,single_code){
  if(geo =="LSOA"){
    ogpURL<-"https://services1.arcgis.com/ESMARspQHYMw9BZ9/arcgis/rest/services/LSOA_Dec_2011_Boundaries_Generalised_Clipped_BGC_EW_V3_2022/FeatureServer/0/query"
    geocode<-"LSOA11CD"
    init<-"OA"
    }
  if(geo =="OA"){
    ogpURL<-"https://services1.arcgis.com/ESMARspQHYMw9BZ9/arcgis/rest/services/Output_Areas_Dec_2011_Boundaries_EW_BGC_2022/FeatureServer/0/query?"
    geocode<-"OA11CD"
    init<-"LS"
  }

  if(geo =="WZ"){
    ogpURL<-"https://services1.arcgis.com/ESMARspQHYMw9BZ9/arcgis/rest/services/Workplace_Zones_Dec_2011_FCB_in_England_and_Wales_2022/FeatureServer/0/query"
    geocode<-"wz11cd"
    init<-"OA"
  }


  cd<-data %>%
    dplyr::select(!tidyselect::starts_with(init)) %>%
    dplyr::select(tidyselect::contains(geocode)) %>%
    as.data.frame()

  if(nrow(cd)==1){
    #whereClause<-urltools::url_encode(paste0("?where=",colnames(cd),cd[,1]))
    whereClause<-paste0("?where=",colnames(cd),"%20%3D%20'",cd[,1],"'")
    ogpURL<-paste0(ogpURL,whereClause,"&outFields=*&outSR=4326&f=json")
    sf <- sf::st_read(ogpURL)
  }else{
    seq<-splitAt(cd[,1],50)
    clause<-seq %>% purrr::map(~paste(.x,sep = "",collapse = "', '"))
    clause<-clause %>% purrr::map(~paste0(geocode," IN ('",noquote(.x),"')"))
    req <- clause %>% purrr::map(~ httr::POST(url = ogpURL,
                                              body = list(where= .x,
                                              outfields="*",
                                              outSR = '4326',
                                              f='json'),
                                .headers = c("application/x-www-form-urlencoded"))
                                )
    sf <- req %>% purrr::map_df(~sf::read_sf(httr::content(.x,type='text',encoding='UTF-8')))
  }

  sf <- sf[c("geometry")]
  data<-sf::st_as_sf(cbind(data,sf))
  return(data)
}

splitAt <- function(x, pos) {
  unname(split(x,rep(c(0:(round(length(x)/pos))),each=pos)[1:length(x)]))
}

search_ladCode<-function(name){
  urlLAD<-"https://services1.arcgis.com/ESMARspQHYMw9BZ9/arcgis/rest/services/LAD_DEC_2020_UK_BGC/FeatureServer/0/query?"
  if(length(name)==1){
      urlLAD<-paste0(urlLAD,"where=LAD20NM%20%3D%20'",name,"'&outFields=*&outSR=4326&f=json")
      lad<-sf::st_read(urlLAD)
      ldCD<-lad$LAD20CD
  }else{
    clause<-paste(name,sep = "",collapse = "', '")
    clause<-paste0("LAD20NM IN ('",clause,"')")
    req<-httr::POST(url = urlLAD,
               body = list(where= clause,
                           outfields="*",
                           outSR = '4326',
                           f='json'),
               .headers = c("application/x-www-form-urlencoded"))
    lads<-sf::read_sf(httr::content(req,type='text',encoding='UTF-8'))
    ldCD<-lads$LAD20CD
  }

  return(ldCD)
}




